Feat/cross encoder trainer listnet

milistu opened the pull request:
# ListNet loss Implementation for Cross Encoder Trainer

This PR adds ListNet loss functionality to the Cross Encoder Trainer feature.

## Changes

- Implemented ListNet loss as a listwise loss function
   - Add activation function argument in loss implementation
- Added support for valid label columns and dataclass decorator - `sentence_transformers/cross_encoder/data_collator.py`
- Created an example script demonstrating ListNetLoss usage

## Implementation Details

ListNetLoss is a listwise loss function that uses a probability model for ranked lists. It's particularly effective for learning to rank tasks in information retrieval systems by directly optimizing for the entire list of documents rather than pairs.

Reference: https://www.microsoft.com/en-us/research/wp-content/uploads/2016/02/tr-2007-40.pdf

milistu committed:
Add new valid lael columns, add dataclass decorator.

milistu committed:
Add ListNetLoss.

milistu committed:
Add ListNetLoss, new listwise loss.

milistu committed:
Add example script for ListNetLoss.

milistu committed:
Change model, add separate eval batch.

milistu committed:
Add activation function argument.

tomaarsen commented on sentence_transformers/cross_encoder/losses/ListNetLoss.py:
```diff
@@ -0,0 +1,148 @@
+        model: CrossEncoder,
+        eps: float = 1e-10,
+        pad_value: int = -1,
+        activation_fct: nn.Module | None = nn.Sigmoid(),
```
We got better results with `nn.Identity` here right? Whatever worked best for this particular loss should be the default

milistu replied:
Good catch, my apologies for the oversight. I've updated the default to `Identity`, since that gave us better results for this loss

tomaarsen committed:
Update metric_for_best_model name

I updated the name in NanoBEIR to include R100 for reranking the top 100 from BM25.

tomaarsen committed:
Ensure self.activation_fct is always nn.Module

tomaarsen committed:
Use fullname(self.activation_fct) in config dict, instead of the instance itself

tomaarsen committed:
Set requires_grad for 0.0 loss edge case

tomaarsen committed:
Replace manual Cross Entropy with nn.CrossEntropyLoss

I verified that this gives identical loss scores

tomaarsen committed:
Call the activation function on logits, not full_logits, and avoid -inf

-inf is problematic with CrossEntropyLoss with class probabilities rather than class weights: becomes nan

tomaarsen committed:
Merge branch 'feat/cross_encoder_trainer' of https://github.com/tomaarsen/sentence-transformers into tom-pr-5

milistu committed:
Replace default activation function from Sigmoid to Identy.

tomaarsen committed:
Remove now-unused eps

This is not needed for nn.CrossEntropyLoss

tomaarsen commented:
I'm rerunning the training script with both `identity` and `sigmoid` to make sure :)

If I understand correctly, then the only thing that's in our way to prevent variable-length document lists is the data collator that can't turn the `labels` into a tensor?
```python
  File "[...]\sentence-transformers\sentence_transformers\cross_encoder\data_collator.py", line 45, in __call__
    batch["label"] = torch.tensor([row[label_column] for row in features])
                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
ValueError: expected sequence of length 2 at dim 1 (got 3)
```

We could simply remove the `torch.tensor` call, and just pass a list of integers/floats/whatever to the loss? Right now, i.e. before we push out the first release of this, is the only time we can make a change like this. The variable shapes can be handled on the loss level (or the loss can choose to not allow them). What do you think?

Oh, another thing that I was considering: What if we update it such that `logits = self.model(**tokens)[0].view(-1)` is computed in batches of `batch_size` pairs at once? Perhaps that's a bit more "consistent" for the GPU, and perhaps faster? I'm noticing there's a decent bit of spikes in the GPU usage, but that could also just be because it has to tokenize in the loss itself as it can't do it beforehand.

---

I'm working on a bunch of small changes based on this PR now: i.e. using `nn.CrossEntropyLoss` instead of manual cross-entropy because (my understanding is that) it uses a logsumexp trick to prevent numerical instability. In my tests, it gave identical losses. This one can then also (potentially) be overridden with a FusedCrossEntropyLoss (e.g. [here](https://olmo-core.readthedocs.io/en/latest/nn/functional.html#olmo_core.nn.functional.fused_cross_entropy_loss)), which has been all the rage recently.

I'm also updating the docs so it's all fully integrated with the rest 🤗 

Please have a look at my changes, I'm open to suggestions!

- Tom Aarsen

tomaarsen committed:
Merge branch 'feat/cross_encoder_trainer_listnet' of https://github.com/milistu/sentence-transformers into tom-pr-5

tomaarsen committed:
Add ListNetLoss to docs fully

tomaarsen committed:
Mention "label", "labels", "score", or "scores" in training overview

milistu commented:
> I'm rerunning the training script with both `identity` and `sigmoid` to make sure :)

Definitely, based only on my omission where I applied `Softmax` on full logits, it raises a concern about my findings. Please do share findings after all the changes that you applied. Also, I was fully focused on integrating loss from paper that I completely forgot to use `CrossEntropy` from torch...

> If I understand correctly, then the only thing that's in our way to prevent variable-length document lists is the data collator that can't turn the `labels` into a tensor?
> 
> ```python
>   File "[...]\sentence-transformers\sentence_transformers\cross_encoder\data_collator.py", line 45, in __call__
>     batch["label"] = torch.tensor([row[label_column] for row in features])
>                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> ValueError: expected sequence of length 2 at dim 1 (got 3)
> ```
> 
> We could simply remove the `torch.tensor` call, and just pass a list of integers/floats/whatever to the loss? Right now, i.e. before we push out the first release of this, is the only time we can make a change like this. The variable shapes can be handled on the loss level (or the loss can choose to not allow them). What do you think?

Let me first check if I got this correct.
You want to support variable lengths directly without the need for data preprocessing where we expand labels and documents. This would definitely make things easier. I was exploring this option initially but assumed it would require a big architectural change, so I implemented a workaround with dataset preprocessing instead.

If so, I want to brainstorm a little to see what would happen inside loss:
- We would have a main `batch_size` for the number of query-documents pairs that are being passed to the loss function
- Next we would need a separate `batch_size` for CE processing of query-doc pairs - This would make GPU usage more stable. There would probably be a last batch with a different number of pairs, but I think this won't be a problem.
- After that, based on the original inputs, we can arrange the logits to match the input shape
- Then we can apply an activation function
- At the end, expand the logits with -inf and labels so we can calculate loss in parallel for different examples

Did I get the logic right? What do you think?
I agree that this is the time to make this change, later will be a big change that won't be that easy to implement.

> Oh, another thing that I was considering: What if we update it such that `logits = self.model(**tokens)[0].view(-1)` is computed in batches of `batch_size` pairs at once? Perhaps that's a bit more "consistent" for the GPU, and perhaps faster? I'm noticing there's a decent bit of spikes in the GPU usage, but that could also just be because it has to tokenize in the loss itself as it can't do it beforehand.

My initial idea for padding was to add an empty string for missing documents and pass that to the CE. Technically this would make GPU utilization stable but we would waste resources. I immediately knew that this idea was bad, but it was good enough for testing. Later I added cleaning where I take only existing pairs and pass them to CE:
```python
mask = labels != self.pad_value  # shape: (batch_size, max_docs)

        batch_indices, doc_indices = torch.where(mask)

        pairs = [
            (queries[batch_index], docs_list[batch_index][doc_index])
            for batch_index, doc_index in zip(batch_indices.tolist(), doc_indices.tolist())
        ]
```
I am positive that this part is making GPU spikes. Because at every batch we can get a different amount of real (not padded) values and when passing to CE for logic computation, there is a different GPU utilization. Adding batch to this part would definitely make the GPU more stable. If we decide to go that route, that would mean that we need to add another argument for `batch_size` inside loss, or do you have a better solution?

> I'm working on a bunch of small changes based on this PR now: i.e. using `nn.CrossEntropyLoss` instead of manual cross-entropy because (my understanding is that) it uses a logsumexp trick to prevent numerical instability. In my tests, it gave identical losses. This one can then also (potentially) be overridden with a FusedCrossEntropyLoss (e.g. [here](https://olmo-core.readthedocs.io/en/latest/nn/functional.html#olmo_core.nn.functional.fused_cross_entropy_loss)), which has been all the rage recently.

This is much better, cleaner code, and I would not be surprised to get a better result at the end compared to my implementation. To be honest, I was not familiar with `FusedCrossEntropyLoss` but from what I read I see that improvements are promising and we should definitely test this out 🧪

> I'm also updating the docs so it's all fully integrated with the rest 🤗

Thank you 🙏, I left the docs for last because I was not sure if the losses would be accepted 😄

> Please have a look at my changes, I'm open to suggestions!

I agree with all of your changes!

tomaarsen commented:
Regarding my Identity and Sigmoid tests, here is the distribution of outputs after inference. Inference is done with Sigmoid for both. Both models also performed roughly equivalently on my benchmarks:
* https://huggingface.co/tomaarsen/reranker-msmarco-v1.1-MiniLM-L12-H384-uncased-listnet-sigmoid
* https://huggingface.co/tomaarsen/reranker-msmarco-v1.1-MiniLM-L12-H384-uncased-listnet-identity

![image](https://github.com/user-attachments/assets/9c8a0bd7-4a2b-4b88-a0e0-c503938ab42c)

In the end, both seem fine. It's all about the ranking anyways, not really the absolute scores, but I'd just like to avoid models exclusively giving either 0.0 or 1.0.

tomaarsen commented:
@milistu Apologies, I totally missed your message when I posted my Identity/Sigmoid test results.

1. Yes, your logic seems solid. This is what I had planned as well.
2. If we compute the forwards in batches, we can indeed introduce a `mini_batch_size` argument. This is also used in `CachedMultipleNegativesRankingLoss`, so it should be semi-familiar. Like in CMNRL, the `mini_batch_size` only informs the memory usage/speed, while `per_device_train_batch_size` from TrainingArguments determines the size of a batch (which is relevant for learning rates/how well the model trains).

For ListNetLoss (and LambdaLoss?) we can also keep it `None` by default, and then set it as `batch_size` (a.k.a. `len(inputs)`) if it's not defined.

I will work on this today, if that's okay with you!

- Tom Aarsen

tomaarsen committed:
Allow variable docs per query; batchify ListNetLoss

tomaarsen commented:
Done!
I trained a model with it here: https://huggingface.co/tomaarsen/reranker-msmarco-v1.1-MiniLM-L12-H384-uncased-listnet
(Note, performance is a bit down compared to the previous (https://huggingface.co/tomaarsen/reranker-msmarco-v1.1-MiniLM-L12-H384-uncased-listnet-original), likely because this is 1 epoch instead of 3)

Please have a look at it - feel free to make any suggestions, I changed quite a bit to get this working 🤗 

- Tom Aarsen

milistu commented:
> Done! I trained a model with it here: https://huggingface.co/tomaarsen/reranker-msmarco-v1.1-MiniLM-L12-H384-uncased-listnet (Note, performance is a bit down compared to the previous (https://huggingface.co/tomaarsen/reranker-msmarco-v1.1-MiniLM-L12-H384-uncased-listnet-original), likely because this is 1 epoch instead of 3)
> 
> Please have a look at it - feel free to make any suggestions, I changed quite a bit to get this working 🤗
> 
> * Tom Aarsen

@tomaarsen 

This is a **huge** improvement, nice work and **thank you**, Tom! 
I'll update LambdaLoss with similar changes and dive into all the details to expand my understanding further 😄

tomaarsen commented:
Perfect! I've merged this so you can pull the data collator etc. changes into your LambdaLoss PR.

Thanks a bunch for setting this all up! I think it's very valuable.

- Tom Aarsen
